File: packages/volt-cli/src/Volt.Cli.Sync/IdeTree.cs
namespace Volt.Cli.Sync;

/// <summary>
/// refs/remotes/volt/ide — the live IDE modelled as a git remote-tracking branch. Each commit's tree is the
/// user's branch tree with ONLY <c>src/</c> swapped for the IDE's state, so the merge never touches the scaffold.
/// C# port of packages/volt-git/src/domain/ide-tree.ts — the correctness-critical merge engine.
/// </summary>
public static class IdeTree
{
    public const string Range = "refs/remotes/volt/ide";

    public static string? VoltIdeHead(string gitDir) => Git.ResolveRef(gitDir, Range);

    /// <summary>
    /// Build the volt/ide tree = the IDE's current state under <c>src/</c>, plus the user's scaffold from HEAD.
    /// Tracked src/ files: changed items → the fetched content; unchanged items → carried from the PARENT volt/ide
    /// tree (the IDE's last-known content, NOT HEAD — HEAD holds the user's un-pushed edits, and sourcing an
    /// unchanged item from HEAD would fold that edit into the IDE baseline and strand it forever); removed items →
    /// dropped. Everything else (non-src scaffold, non-tracked src files) is carried from HEAD (the user's side),
    /// so the merge only ever touches the IDE axis. On init parentIde is null and ideFiles is the whole IDE.
    /// </summary>
    public static string BuildVoltIdeTree(
        string gitDir,
        string? headCommit,
        string? parentIde,
        IReadOnlyList<MaterializedFile> ideFiles,
        IReadOnlyList<string> removedNames)
    {
        var entries = new List<IndexEntry>();
        var seen = new HashSet<string>();
        void Add(IndexEntry e) { if (seen.Add(e.Path)) entries.Add(e); }

        string? SrcRel(string path) => path.StartsWith(Files.SrcDir + "/", StringComparison.Ordinal) ? path.Substring(Files.SrcDir.Length + 1) : null;
        var replaced = new HashSet<string>(ideFiles.Select(f => f.Path));
        var removed = new HashSet<string>(removedNames);

        // Changed IDE items — fresh content from the fetch.
        foreach (var f in ideFiles)
            Add(new IndexEntry("100644", Git.WriteBlob(gitDir, f.Content), $"{Files.SrcDir}/{f.Path}"));

        // Unchanged IDE items — from the previous volt/ide tree (the IDE's last-known content, NOT the user's HEAD).
        if (parentIde is not null)
            foreach (var e in Git.ListTree(gitDir, parentIde))
            {
                var rel = SrcRel(e.Path);
                if (rel is not null && Extensions.IsTrackedPath(rel) && !replaced.Contains(rel) && !removed.Contains(rel))
                    Add(new IndexEntry(e.Mode, e.Sha, e.Path));
            }

        // Scaffold + non-tracked src/ files — from HEAD (the user's side; the merge leaves these untouched).
        if (headCommit is not null)
            foreach (var e in Git.ListTree(gitDir, headCommit))
            {
                var rel = SrcRel(e.Path);
                if (rel is null || !Extensions.IsTrackedPath(rel))
                    Add(new IndexEntry(e.Mode, e.Sha, e.Path));
            }

        return Git.BuildTree(gitDir, entries);
    }

    public static string CommitVoltIde(string gitDir, string treeSha, string? parent, string message) =>
        Git.CommitTree(gitDir, treeSha, parent is not null ? new[] { parent } : Array.Empty<string>(), message);
}

File: packages/volt-git/src/domain/ide-tree.ts
/**
 * refs/remotes/volt/ide — the live IDE modelled as a git REMOTE-tracking branch, so it shows in the
 * graph as `volt/ide` (the IDE is literally a remote you fetch+merge on pull / push to on push). Each
 * commit's tree is the user's branch tree with ONLY `src/` swapped for the IDE's state, so the merge
 * never touches the scaffold. Living under refs/remotes/ means it's visible locally but never pushed to
 * a real origin. The optimistic-concurrency baseline (what the IDE last had) lives in the sidecar
 * (`sidecar.ts`, `.git/volt/ide-refs.json`).
 */
import { buildTree, commitTree, listTree, resolveRef, writeBlob, type IndexEntry } from "../git.js";
import { isTrackedPath } from "./extensions.js";
import type { MaterializedFile } from "./materialize.js";
import { SRC_DIR } from "../files.js";

export const RANGE = "refs/remotes/volt/ide";

export function voltIdeHead(gitDir: string): string | undefined {
	return resolveRef(gitDir, RANGE);
}

/**
 * Build the volt/ide tree = the IDE's current state under `src/`, plus the user's scaffold from HEAD.
 *
 * Tracked `src/` files (the IDE's items):
 *   - changed items      → the freshly fetched content (`ideFiles`)
 *   - unchanged items    → carried from the PARENT volt/ide tree (`parentIde`) — the IDE's last-known
 *                          content. NOT from HEAD: HEAD holds the USER's edits, and `/fetch` is incremental
 *                          (an item the IDE didn't touch is absent from `ideFiles`), so sourcing it from HEAD
 *                          would fold the user's un-pushed edit into the IDE baseline and strand it forever
 *                          (it would never diff as outgoing). A user-added item absent from `parentIde` is
 *                          correctly left out here, so it surfaces as outgoing and pushes to the IDE.
 *   - removed items      → dropped (`removedNames`)
 * Everything else — non-`src/` scaffold and any non-tracked `src/` file — is carried from HEAD (the user's
 * side), so the merge only ever touches the IDE axis. On init `parentIde` is undefined and `ideFiles` is the
 * whole IDE, so the parent loop is simply empty.
 */
export function buildVoltIdeTree(
	gitDir: string,
	headCommit: string | undefined,
	parentIde: string | undefined,
	ideFiles: readonly MaterializedFile[],
	removedNames: readonly string[],
): string {
	const entries: IndexEntry[] = [];
	const seen = new Set<string>();
	const add = (e: IndexEntry): void => {
		if (!seen.has(e.path)) {
			seen.add(e.path);
			entries.push(e);
		}
	};
	const srcRel = (path: string): string | null => (path.startsWith(`${SRC_DIR}/`) ? path.slice(SRC_DIR.length + 1) : null);
	const replaced = new Set(ideFiles.map((f) => f.path));
	const removed = new Set(removedNames);

	// Changed IDE items — fresh content from the fetch.
	for (const f of ideFiles) add({ mode: "100644", sha: writeBlob(gitDir, f.content), path: `${SRC_DIR}/${f.path}` });

	// Unchanged IDE items — from the previous volt/ide tree (the IDE's last-known content, not the user's HEAD).
	if (parentIde !== undefined) {
		for (const e of listTree(gitDir, parentIde)) {
			const rel = srcRel(e.path);
			if (rel !== null && isTrackedPath(rel) && !replaced.has(rel) && !removed.has(rel)) add(e);
		}
	}

	// Scaffold + non-tracked src/ files — from HEAD (the user's side; the merge leaves these untouched).
	if (headCommit !== undefined) {
		for (const e of listTree(gitDir, headCommit)) {
			const rel = srcRel(e.path);
			if (rel === null || !isTrackedPath(rel)) add(e);
		}
	}

	return buildTree(gitDir, entries);
}

export function commitVoltIde(gitDir: string, treeSha: string, parent: string | undefined, message: string): string {
	return commitTree(gitDir, treeSha, parent !== undefined ? [parent] : [], message);
}
